added y/n prompt

MCWertGaming opened the pull request:
This PR adds some variations of a y/n style prompt.

Todo:
- [x] unify blocking + non blocking functions
- [x] rename blocking to something else
- [x] use COR
- [x] fix windows compilation

wolfv commented on examples/prompt_non_blocking.cpp:
```diff
@@ -0,0 +1,41 @@
+              << CPP_TERMINAL_VERSION_COMPLETE << std::endl;
+    try {
+        std::cout << "CPP-Terminal basic prompt example: \n\n";
+        switch (Term::prompt_non_blocking("Proceed?", "Y", "n", ":")) {
```
is this blocking in the sense of "blocking call" or wether `Enter` has to be pressed or not?

If the latter, I think another name might be better. 

It might be good to have a "non-blocking" prompt, too, so that one could cancel after some timeout or similar.

MCWertGaming replied:
I have created notes in the `cpp-terminal/prompt.hpp` file on that. Basically "blocking" is that the the function is waiting for the user to hit the enter key - like i a normal console. The non-blocking version is immediately returning after the first key is pushed.

Maybe we should rename the function into something better describing...
I'd suggest to add another argument for adding a timeout for the function.

wolfv replied:
I think if I read "blocking" or "non-blocking" I would think that the program can continue while the prompt is shown (regardless wether Enter is required or not). So yeah, maybe we can find a better name...

MCWertGaming replied:
Any idea for a name? Will add a simple timeout after I fixed the windows CI.

wolfv replied:
something like `immediate_prompt`? or `prompt_need_confirm` or `prompt_enter` or one could also make it an argument. 
I also quite like `structs` as containers for options, such as 

```
struct prompt_options {
bool needs_enter = false;
bool multiline = false;
bool simple = true; ...
}
```

wolfv replied:
Another way is via flags, such as `NEED_ENTER | MULTILINE`, and then each of those is an integer with a bit flag. 

But you guys need to decide that :)

certik replied:
Perhaps the easiest is just boolean arguments, possibly with default values.

MCWertGaming replied:
adding an option for the multi-line prompt isn't that of of a good i idea I would say. It's really complicated. I'll add a bool for toggling the confirmation. I think that a struct would be more complicated than just passing 5/6 short arguments.

wolfv commented:
Looks really good! I left some questions, I hope you don't mind!

MCWertGaming commented on examples/prompt_simple.cpp:
```diff
@@ -0,0 +1,30 @@
+    try {
+        std::cout << "CPP-Terminal basic prompt example: \n\n";
+        switch (Term::prompt_simple("Proceed?")) {
+            case Term::Result_simple::YES:
```
I have made with the same subset at first, but I thought that this way makes the return more easily to understand and implement. Especially because the simple version is just wrapping the blocking one with default values and already handled default options.

I can change that though, if you prefer it to be one. In the end it's just a matter of preference after all i guess.

MCWertGaming commented:
This needs some work anyway - MSVC is going crazy out of nowhere just because I included the `base.hpp` file that includes the `platform.hpp` file. I think it's something window.h related again or something... Not sure, if I can fix it that easily for now. We have to remove that include anyway to end stuff like this. But I'd suggest to sinish this PR and merge it until I fixed the windows problem seperately.

MCWertGaming commented:
@wolfv do you guys want the prompt to read on non-interactive terminals to read input from the stdin? a use case would be:

```
program | yes # <- send input via standard in
```
It's already implemented (in theory) in CPP-Terminal, I'd just need to make it available to the prompt function.

wolfv commented:
Hmm, good question. I think it woudl be nice, yeah, but most of the time in non-interactive mode we just use `--always-yes`

certik commented on cpp-terminal/prompt.hpp:
```diff
@@ -4,6 +4,42 @@
+Result prompt_blocking(std::string message,
+                       std::string first_option,
+                       std::string second_option,
+                       std::string prompt_indicator);
```
I think you want to do `const std::string &message` and the same for all the other arguments also.

MCWertGaming replied:
The problem with copy by reference is that you cant pass a value directly then, like

```
void test(std::string test);

void test_ref(std::string& test);

int main() {

// without COR
test("STRING");
std::string test_s = "test";
test(test_s);

// with COR
test_ref("STRING");                  // <- This will fail because the const char* is not a variable that can be referenced
test_ref(test_s);                        // <- This works because the variable can be referenced
}
```

But i'll definitely check on that when I added the missing features.

certik replied:
> The problem with copy by reference is that you cant pass a value directly

You can. Here is an example:
```c++
#include <string>

void f1(std::string x) {
}

void f2(const std::string &x) {
}

int main() {
    f1("ok");
    f2("ok");
    return 0;
}
```
Notice the `const`. If you don't put `const`, then indeed you get a compiler error.

MCWertGaming replied:
Oh, i didn't knew that const makes such a difference here. Will do that!

MCWertGaming commented:
I have tested some fixes for the windows CI in #125 and removing the "platform.hpp" include in the "base.hpp" file fixes all errors (of course that is not a possible fix, as we need it to enable the "raw mode" for capturing input). I'll have to make some bigger changes for fixing this. Originally i planed on fixing that after having the prompt done, but well...

MCWertGaming commented:
This PR is now ready @certik @wolfv. Please review it. I have fixed the windows error with a simple undefine for now. It looks ugly, but i'll have to finish the Platform header in order to fix that properly (which will be my next PR). I have moved adding support for reading the choice from stdin to #146.

MCWertGaming commented:
I have added CTRL+D support. Can you guys test this PR?

MCWertGaming commented:
@certik @wolfv what is the state on this PR?

MCWertGaming commented:
It's been almost a month now. Can you review this PR? @certik @wolfv.

MCWertGaming commented:
So what's the status of this PR?

MCWertGaming commented:
@certik @wolfv

certik commented:
Let me review this.

certik commented:
It looks fine to me. @wolfv do you think this goes in the right direction? We can always improve upon it once it is merged.

MCWertGaming commented:
It has been two weeks now. @wolfv please review.

MCWertGaming commented:
What about this PR now? @certik @wolfv I would like to merge the other PRs as well some day.

certik reviewed (approved):
I think that this is fine.

Overall I don't like including windows header files in our header files, they should be included in cpp file only.

I suggest we merge this and iterate on it in master. And if improvements are needed, we can add them in later.
